integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/SparkSQLExecutionContext.java

```diff
@@ -66,249 +42,154 @@
     log.debug("SparkListenerSQLExecutionStart - executionId: {}", startEvent.executionId());
-    startEvent(startEvent.time());
+    if (!olContext.getQueryExecution().isPresent()) {
+      log.info("No execution info {}", olContext);
+      return;
+    }
+    RunEvent event =
+        runEventBuilder.buildRun(
+            buildParentFacet(),
+            openLineage.newRunEventBuilder().eventTime(toZonedTime(startEvent.time())),
+            buildJob(olContext.getQueryExecution().get()),
+            startEvent);
+
+    log.debug("Posting event for start {}: {}", executionId, event);
+    eventEmitter.emit(event);
   }
 
   public void end(SparkListenerSQLExecutionEnd endEvent) {
     log.debug("SparkListenerSQLExecutionEnd - executionId: {}", endEvent.executionId());
     // TODO: can we get failed event here?
     // If not, then we probably need to use this only for LogicalPlans that emit no Job events.
     // Maybe use QueryExecutionListener?
-    endEvent(endEvent.time(), "COMPLETE", null);
-  }
-
-  @Override
-  public void setActiveJob(ActiveJob activeJob) {}
+    if (!olContext.getQueryExecution().isPresent()) {
+      log.info("No execution info {}", olContext);
+      return;
+    }
+    RunEvent event =
+        runEventBuilder.buildRun(
+            buildParentFacet(),
+            openLineage.newRunEventBuilder().eventTime(toZonedTime(endEvent.time())),
+            buildJob(olContext.getQueryExecution().get()),
+            endEvent);
 
-  @Override
-  public void start(SparkListenerJobStart jobStart) {
-    log.debug("SparkListenerJobStart - executionId: " + executionId);
-    jobId = Optional.of(jobStart.jobId());
-    startEvent(jobStart.time());
+    log.debug("Posting event for end {}: {}", executionId, event);
+    eventEmitter.emit(event);
   }
 
   @Override
-  public void end(SparkListenerJobEnd jobEnd) {
-    log.debug("SparkListenerJobEnd - executionId: " + executionId);
-    Exception exception = null;
-    if (jobEnd.jobResult() instanceof JobFailed) {
-      exception = ((JobFailed) jobEnd.jobResult()).exception();
-    }
-    jobId = Optional.of(jobEnd.jobId());
-    endEvent(jobEnd.time(), getEventType(jobEnd.jobResult()), exception);
-  }
-
-  void startEvent(Long time) {
-    if (!started.compareAndSet(false, true)) {
-      log.debug("Start event already emitted: returning");
+  public void start(SparkListenerStageSubmitted stageSubmitted) {
+    if (!olContext.getQueryExecution().isPresent()) {
+      log.info("No execution info {}", olContext);
       return;
     }
+    RunEvent event =
+        runEventBuilder.buildRun(
+            buildParentFacet(),
+            openLineage.newRunEventBuilder().eventTime(ZonedDateTime.now(ZoneOffset.UTC)),
+            buildJob(olContext.getQueryExecution().get()),
+            stageSubmitted);
+
+    log.debug("Posting event for stage submitted {}: {}", executionId, event);
+    eventEmitter.emit(event);
```

Comment:
Would be document when we're emitting events and what particular information we're getting from particular event. It seems unclear to me that we're emitting on `SparkListenerStageSubmitted` and emit is commented for `SparkListenerStageCompleted`.

Reply:
Looks like those aren't even being called right now?

Reply:
They're not - I didn't want to change the current behavior, but I think it would be good to add this behavior in a new PR.

Reply:
I've added a comment and opened https://github.com/OpenLineage/OpenLineage/issues/470